File: infrastructure/laser/gaussian_source.py
"""
Gaussian Laser Source — TEM00 beam profile.

Implements ILaserSource using the standard Gaussian intensity formula:
    I(x,y) = (2P / πw²) · exp(-2r²/w²)

Single Responsibility: Only computes Gaussian intensity distribution.
"""

import numpy as np

from domain.ports import ILaserSource


class GaussianLaserSource(ILaserSource):
    """Strategy: Gaussian (TEM00) beam intensity profile."""

    def compute_intensity(
        self,
        x_grid: np.ndarray,
        y_grid: np.ndarray,
        center_x: float,
        center_y: float,
        spot_radius: float,
        power: float,
    ) -> np.ndarray:
        """
        Compute 2D Gaussian intensity distribution.

        I(x,y) = (2P / πw²) · exp(-2r²/w²)
        """
        r_squared = (x_grid - center_x) ** 2 + (y_grid - center_y) ** 2
        peak_intensity = 2.0 * power / (np.pi * spot_radius ** 2)
        return peak_intensity * np.exp(-2.0 * r_squared / spot_radius ** 2)

File: UnityHVOF/Assets/Scripts/Infrastructure/Laser/GaussianSource.cs
// ============================================================================
// Gaussian Laser Source — Implements ILaserSource.
//
// Pattern: Strategy — interchangeable with TopHatSource.
// Computes: I(x,y) = (2P / πw²) · exp(-2((x-cx)² + (y-cy)²) / w²)
//
// Port of: infrastructure/laser/gaussian_source.py
// ============================================================================

using System;
using HVOFSim.Domain.Ports;

namespace HVOFSim.Infrastructure.Laser
{
    /// <summary>Gaussian (TEM00) beam profile intensity distribution.</summary>
    public sealed class GaussianSource : ILaserSource
    {
        public double[,] ComputeIntensity(
            double[,] xGrid, double[,] yGrid,
            double centerX, double centerY,
            double spotRadius, double power)
        {
            int nx = xGrid.GetLength(0);
            int ny = xGrid.GetLength(1);
            var intensity = new double[nx, ny];

            double prefactor = 2.0 * power / (Math.PI * spotRadius * spotRadius);
            double invW2 = -2.0 / (spotRadius * spotRadius);

            for (int i = 0; i < nx; i++)
            {
                for (int j = 0; j < ny; j++)
                {
                    double dx = xGrid[i, j] - centerX;
                    double dy = yGrid[i, j] - centerY;
                    double r2 = dx * dx + dy * dy;
                    intensity[i, j] = prefactor * Math.Exp(invW2 * r2);
                }
            }

            return intensity;
        }
    }
}
